Occasional build failures in Operationprocessor.testWithDataCorruptionFailures.

**Problem description**
This happens: 
```
io.pravega.segmentstore.server.logs.OperationProcessorTests > testWithDataCorruptionFailures FAILED
    java.lang.AssertionError: Unexpected exception for failed Operation in the same DataFrame as intentionally failed operation. Exception thrown was of unexpected type: java.lang.AssertionError: No Metadata UpdateTransaction found for LastFullySerializedSN = 283, LastStartedSN = 284, Address = Sequence = 16384, Length = 8192
        at io.pravega.test.common.AssertExtensions.assertThrows(AssertExtensions.java:73)
        at io.pravega.segmentstore.server.logs.OperationProcessorTests.testWithDataCorruptionFailures(OperationProcessorTests.java:434)

        Caused by:
        java.util.concurrent.CompletionException: java.lang.AssertionError: No Metadata UpdateTransaction found for LastFullySerializedSN = 283, LastStartedSN = 284, Address = Sequence = 16384, Length = 8192
            at java.util.concurrent.CompletableFuture.reportJoin(CompletableFuture.java:375)
            at java.util.concurrent.CompletableFuture.join(CompletableFuture.java:1934)
            at io.pravega.test.common.AssertExtensions.assertThrows(AssertExtensions.java:69)
            ... 1 more

            Caused by:
            java.lang.AssertionError: No Metadata UpdateTransaction found for LastFullySerializedSN = 283, LastStartedSN = 284, Address = Sequence = 16384, Length = 8192
```

**Problem location**
OperationProcessor.State.

**Suggestions for an improvement**

--- a/segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/DataFrameBuilder.java
+++ b/segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/DataFrameBuilder.java
@@ -287,8 +287,8 @@ class DataFrameBuilder<T extends SequencedItemList.Element> implements AutoClose
 
         @Override
         public String toString() {
-            return String.format("LastFullySerializedSN = %d, LastStartedSN = %d, Address = %s, Length = %d",
-                    getLastFullySerializedSequenceNumber(), getLastStartedSequenceNumber(), this.logAddress, getDataFrameLength());
+            return String.format("TxnId = %d, LastFullySerializedSN = %d, LastStartedSN = %d, Address = %s, Length = %d",
+                    getMetadataTransactionId(), getLastFullySerializedSequenceNumber(), getLastStartedSequenceNumber(), this.logAddress, getDataFrameLength());
         }
     }
 

--- a/segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/OperationProcessor.java
+++ b/segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/OperationProcessor.java
@@ -645,7 +645,17 @@ class OperationProcessor extends AbstractThreadPoolService implements AutoClosea
                 }
             }
 
-            assert checkpointExists : "No Metadata UpdateTransaction found for " + commitArgs;
+            if (!checkpointExists) {
+                // Under normal circumstances, there should always be an UpdateTransaction that matches our argument;
+                // however if we had just processed a failure, collectFailureCandidates() may have cleared (and failed)
+                // all of them, so, only in that case, would it be OK not to find one.
+                log.warn("{}: No Metadata UpdateTransaction found for '{}' (Count={}). This is expected after a critical failure or when OperationProcessor is shutting down.",
+                        traceObjectId, this.metadataTransactions.size(), commitArgs);
+
+                // If a failure did happen, then there should be no other entries in metadataTransactions at this point.
+                assert this.metadataTransactions.isEmpty() : "No Metadata UpdateTransaction found for given CommitArgs, "
+                        + "but there are still entries in metadataTransaction.";
+            }
             return toAck;
         }
 

--- a/segmentstore/server/src/test/java/io/pravega/segmentstore/server/logs/OperationProcessorTests.java
+++ b/segmentstore/server/src/test/java/io/pravega/segmentstore/server/logs/OperationProcessorTests.java
@@ -374,7 +374,6 @@ public class OperationProcessorTests extends OperationLogTestBase {
      * is generated.
      */
     @Test
-    @SuppressWarnings("checkstyle:CyclomaticComplexity")
     public void testWithDataCorruptionFailures() throws Exception {
         // If a DataCorruptionException is thrown for a particular Operation, the OperationQueueProcessor should
         // immediately shut down and stop accepting other ops.

